pycobertura/reporters.py

```diff
@@ -14,77 +14,54 @@
 file_row = namedtuple("FileRow", row_attributes)
 file_row_missed = namedtuple("FileRowMissed", row_attributes + ["missed_lines"])
 
+headers_without_missing = ["Filename", "Stmts", "Miss", "Cover"]
+headers_missing = ["Filename", "Stmts", "Miss", "Cover", "Missing"]
+
 
 class Reporter(object):
     def __init__(self, cobertura):
         self.cobertura = cobertura
 
+    @staticmethod
+    def format_line_rate(line_rate):
+        return f"{line_rate:.2%}"
+
     def get_report_lines(self):
-        lines = []
-
-        for filename in self.cobertura.files():
-            row = file_row_missed(
-                filename,
-                self.cobertura.total_statements(filename),
-                self.cobertura.total_misses(filename),
-                self.cobertura.line_rate(filename),
-                self.cobertura.missed_lines(filename),
-            )
-            lines.append(row)
-
-        footer = file_row_missed(
-            "TOTAL",
-            self.cobertura.total_statements(),
-            self.cobertura.total_misses(),
-            self.cobertura.line_rate(),
-            [],  # dummy missed lines
-        )
-        lines.append(footer)
+        lines = {
+            "Filename": [filename for filename in self.cobertura.files()],
+            "Stmts": [
+                self.cobertura.total_statements(filename)
+                for filename in self.cobertura.files()
+            ],
+            "Miss": [
+                self.cobertura.total_misses(filename)
+                for filename in self.cobertura.files()
+            ],
```

Comment:
I'm trying to understand the performance implications of iterating over all filenames as many times as we have arrays to fill. Should we be worried?

I think ultimately we are doing the same amount of operations, but I'm wondering what the overhead cost is of iterating over those file names is – whereas before we did it in one pass.

Reply:
List comprehensions should be fast, but what I did is store `self.cobertura.files()` as Reporter class variable.